scripts/Calc_hru_wgtavg_nc.py

```diff
@@ -1,208 +1,137 @@
-            c += 1
-            if c == 1:
-              oriVal = Val1
+    att_data = getattr(var, att_name)
+    return att_data
+
+
+def comp_agv_val(nc_wgt, nc_in, varname, chunk):
+    """Compute areal weighted avg value of <varname> in <nc_in>
+       for each of <chunk> hrus based on hru's weight in <nc_wgt>
+    """
+    hru_ids = get_hru_id(nc_wgt)
+    hru_data = split(hru_ids, NPROC)
+    hru_list = hru_data[chunk]
+
+    var_data = get_netCDF_data(nc_in, varname)
+    fill_val = get_netCDF_attr(nc_in, varname, '_FillValue')
+    ghruid = get_netCDF_data(nc_in, GRID_ID_NAME)
+    dim1size = var_data.shape[0]
+    # Initialize wgt_ave_val[ntime,nhru]
+    wgt_ave_val = np.full((dim1size, len(hru_list)), FILL_VALUE)
+    for i in range(len(hru_list)):
+        print hru_list[i]  # Get list of wgt for corresponding hru
+        (wgt_array, overlap_ids) = get_wgt(nc_wgt, hru_list[i])
+        # Go through wgt list and replace value with zero for following cases
+        # where overlapping polygon has missing value - case1
+        # where overlapping polygon is outside nc_wgt domain - case2
+        a = np.zeros((dim1size, len(wgt_array)))
+        for j, overlap_id in enumerate(overlap_ids):
+            # find index of grid cell that match up with hru id of overlap_ids
+            row, col = np.where(ghruid == overlap_id)
+            # if nc_in netCDF does not cover nc_wgt domain - case2
+            if not (np.size(row) and np.size(col)):
+                wgt_array[j] = 0.0
             else:
-              oriVal=np.concatenate((oriVal, Val1), axis=1)
-      # This is weighted average value at current hru
-      Val2 = oriVal.sum(axis=1)
-      del oriVal
-
-    #if there is no valid weight values (i.e. all zero)- nothing overlapping for current hru, so missing value
-    else:
-      Val2 = np.ones((dim1size))*-9999.0
-    # store weighted average time series, Val2 in final array wgtVal
-    wgtVal[:,i]= Val2
-    
-    del Val2
-
-  return wgtVal
-
-############################################
-#                Main                      #
-############################################
+                a[:, j] = np.squeeze(var_data[:, row, col])
+                # if value of overlapping polygon is missing data -case1
+                if any(a[:, j] == fill_val):
+                    wgt_array[j] = 0.0
+        sum_wgt = np.sum(wgt_array)
+        if sum_wgt > 0.0:
+            # Adjust weight value if valid weight value (> 0) exist in list
+            wgt_array = wgt_array/sum_wgt
+            wgt_array = np.where(np.isnan(wgt_array), 0, wgt_array)
+            wgt_ave_val[:, i] = np.dot(a, wgt_array.T).T
+
+    return wgt_ave_val
+
+
+def split(hru_list, n):
+    subsize = len(hru_list)/n
+    idx1 = 0
+    hru_data = []
+    for _ in range(n-1):
+        idx2 = idx1+subsize
+        hru_data.append(hru_list[idx1:idx2])
+        idx1 = idx2
+    hru_data.append(hru_list[idx1:])
```

Comment:
This should be refactored to use the range iterator correctly. I think that would look like this but you should test it.

```Python
+def split(hru_list, n):
    subsize = len(hru_list)/n
    hru_data = []
    for idx1 in range(0, n, subsize):
        idx2 = idx1 + subsize
        hru_data.append(hru_list[idx1: idx2])
    hru_data.append(hru_list[idx1:])
```

Reply:
I tested this and some modified versions, but I don't get the same answer (or getting error on index). I need to test this portion more separately.

Reply:
Maybe something like this, but still does not quite produce what I want. You could check too.

```
def split1(some_list, nsplit):
    subsize = int(len(some_list)/nsplit)
    some_data = []
    for idx1 in range(0, len(some_list)-2*subsize, subsize):
        idx2 = idx1 + subsize
        some_data.append(some_list[idx1: idx2])
    some_data.append(some_list[idx2:])
    return some_data
```

maybe for now leave it is.

Reply:
Can you provide an example of what the hru_list looks like? Now that I look at this again, you may be able to get away with something much simpler:

```Python
def split1(l, n):
    """Yield successive n-sized chunks from l."""
    for i in range(0, len(l), n):
        yield l[i:i + n]
```

ref: https://stackoverflow.com/a/312464/1757464

Reply:
```python
n_split = 4
n_list = 16
some_list = list(range(0,n_list))
print(some_list)
```
```
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
```
```python
split_list = split(some_list, n_split)
print(split_list)
```
should be like 
```
[[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11], [12, 13, 14, 15]]
```

i will try stackoverflow example